sdk/iot/core/src/az_iot_hub_client.c

```diff
@@ -1,12 +1,137 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // SPDX-License-Identifier: MIT
 
+#include <stdint.h>
+
 #include "az_iot_hub_client.h"
+#include <az_precondition.h>
+#include <az_result.h>
 #include <az_span.h>
 
 #include <_az_cfg.h>
 
+static const uint8_t hub_client_forward_slash = '/';
+static const uint8_t hub_client_param_separator = '&';
+static const uint8_t hub_client_null_terminate = '\0';
+
+static const az_span hub_client_api_version = AZ_SPAN_LITERAL_FROM_STR("?api-version=2018-06-30");
+
 AZ_NODISCARD az_iot_hub_client_options az_iot_hub_client_options_default()
 {
   return (az_iot_hub_client_options){ .module_id = AZ_SPAN_NULL, .user_agent = AZ_SPAN_NULL };
 }
+
+AZ_NODISCARD az_result az_iot_hub_client_init(
+    az_iot_hub_client* client,
+    az_span iot_hub_hostname,
+    az_span device_id,
+    az_iot_hub_client_options const* options)
+{
+  AZ_PRECONDITION_NOT_NULL(client);
+  AZ_PRECONDITION_VALID_SPAN(iot_hub_hostname, 1, false);
+  AZ_PRECONDITION_VALID_SPAN(device_id, 1, false);
+  AZ_PRECONDITION_NOT_NULL(options);
+
+  client->_internal.iot_hub_hostname = iot_hub_hostname;
+  client->_internal.device_id = device_id;
+  memcpy(&client->_internal.options, options, sizeof(az_iot_hub_client_options));
+
+  return AZ_OK;
+}
+
+//[Format without module id] {iothubhostname}/{device_id}/?api-version=2018-06-30
+//[Format with module id] {iothubhostname}/{device_id}/{module_id}/?api-version=2018-06-30
+AZ_NODISCARD az_result az_iot_hub_client_user_name_get(
+    az_iot_hub_client const* client,
+    az_span mqtt_user_name,
+    az_span* out_mqtt_user_name)
+{
+  AZ_PRECONDITION_NOT_NULL(client);
+  AZ_PRECONDITION_VALID_SPAN(mqtt_user_name, 0, false);
+  AZ_PRECONDITION_NOT_NULL(out_mqtt_user_name);
+
+  const az_span* module_id = &(client->_internal.options.module_id);
+  const az_span* user_agent = &(client->_internal.options.user_agent);
+
+  if (az_span_capacity(mqtt_user_name)
+      < (int32_t)(
+            az_span_length(client->_internal.iot_hub_hostname) + sizeof(hub_client_forward_slash)
+            + az_span_length(client->_internal.device_id) + sizeof(hub_client_forward_slash)
+            + az_span_length(*module_id)
+            + (az_span_length(*module_id) > 0 ? sizeof(hub_client_forward_slash) : 0)
+            + az_span_length(*user_agent)
+            + (az_span_length(*user_agent) > 0 ? sizeof(hub_client_param_separator) : 0)
+            + az_span_length(hub_client_api_version) + sizeof(hub_client_null_terminate)))
+  {
+    return AZ_ERROR_INSUFFICIENT_SPAN_CAPACITY;
```

Comment:
why do we need to do this? Why can az_span_append fail with this error code? I think this was something that Jeff talked about

Reply:
The behavior should be the same without it as discussed at scrum. The overhead is going through memcpy's even though the end result could be a failure.

Reply:
So in the case that we keep this, the following calls would be converted to:
```
assert(az_span_append(....));
```
to eliminate the additional check. Otherwise, if we don't check for the correct size, we would keep the check on each return value with `AZ_RETURN_IF_FAILED()`